recipe/2024-06-private-bicep-registries.md

```diff
@@ -0,0 +1,425 @@
+```
+Username: AWS
+Password: output of `aws ecr get-login-password --region region`
+```
+While there are various container registry providers, each offering multiple authentication methods, by leveraging ORAS, we abstract away the need for provider-specific code to handle authentication. 
+
+ORAS auth client example to authenticate private registry using username and password,
+```
+	repo.Client = &auth.Client{
+		Client: retry.DefaultClient,
+		Credential: auth.StaticCredential("orasregistry.azurecr.io", auth.Credential{
+			Username: "<username>",
+			Password: "<password>",
+		}),
+	}
+	repo.Client = client
+```
+
+#### Authentication using Federated identity
+**Azure**:
+
+Users can leverage Azure Federated Identity to authenticate seamlessly. When Azure Federated Identity is set up on a Kubernetes cluster, it enables secure and managed authentication, allowing workloads running on the cluster to access ACR without needing to manage service principal credentials directly. This approach enhances security by eliminating the need for hardcoded credentials and simplifies the authentication process by relying on Azure Active Directory (AAD) token issuance for the cluster's managed identity.
+
+[Enable Azure Workload Identity on Radius:](https://docs.radapp.io/guides/operations/providers/azure-provider/howto-azure-provider-wi/)
+```
+rad install kubernetes --set global.azureWorkloadIdentity.enabled=true
+```
+or by configuring azure provider with Azure Credential kind `Workload Identity` using `rad init --full` command.
+
+To authenticate ACR using Azure Federated credentials, we need to get the refresh token from ACR, and use it with ORAS auth client. The detailed steps are outlined below:
+
+- Retrieve the Azure Active Directory (AAD) token for the client ID configured with Azure Federated Identity.
+- Get refresh token from ACR by exchanging for the above AAD access token, set the 
+  ```
+  formData := url.Values{
+      "grant_type":   {"refresh_token"},
+      "service":      test.azurecr.io,
+      "tenant":       <tenant-id>,
+      "access_token": <aad-token>,
+      "expires_in": <5 mins in seconds> 
+    }
+    
+    // jsonResponse contains the refresh token from ACR.
+    jsonResponse, err := http.PostForm("https://test.azurecr.io/oauth2/exchange", formData)
+  ```
+- Use `RefreshToken` in ORAS auth client to authenticate private ACR.(Link to ORAS API docs for auth client : https://pkg.go.dev/oras.land/oras-go@v1.2.6/pkg/registry/remote/auth#Credential)
+  ```
+  repo.Client = &auth.Client{
+		Client: retry.DefaultClient,
+		Credential: auth.StaticCredential(test.azurecr.io, auth.Credential{
+			RefreshToken: <refresh-token>,
+		}),
+	}
+  ```
+
+**AWS**:
+
+Similar to Azure Federated Identity, users can leverage AWS IAM Roles for Service Accounts (IRSA) to authenticate seamlessly to Amazon Elastic Container Registry (ECR). Users are expected to provide `rolearn` as secret in secretStore resources referenced in the recipe config.
+
```

Comment:
Same question for AWS - any bespoke client code we have to write? or do their SDKs do what we need?